### PR TITLE
Fixing SyncBN dgrad

### PR DESCRIPTION
**Pull request by jjsjann123**
Previous PR #22248 which provides support for variadic batch size across processes doesn't account the mean_dy/mean_dy_xmu on backward path, which produces wrong dgrad.

### CODE REVIEW

**Comment by jjsjann123** [comment]
I have a repro code and verified the fix on my local machine.
I'll clean the code and add a tiny test.

**Comment by jjsjann123** [comment]
cc'ing @ptrblck

**Comment by jjsjann123** [comment]
Test added, should be good for review now. @ngimel

File: test/distributed/test_distributed.py [hunk 1]
```diff
@@ -2132,6 +2132,63 @@ def test_DistributedDataParallel_SyncBatchNorm_Diff_Input_Sizes_Running_Value(se
         torch.testing.assert_allclose(running_mean, all_input_var.mean(1))
         torch.testing.assert_allclose(running_var, all_input_var.var(1))
 
+    @unittest.skipIf(BACKEND != 'nccl' and BACKEND != 'gloo',
+                     "Only Nccl & Gloo backend support DistributedDataParallel")
+    @skip_if_no_cuda_distributed
+    @skip_if_no_gpu
+    def test_DistributedDataParallel_SyncBatchNorm_Diff_Input_Sizes_gradient(self):
```
Comment (ngimel) [inline test/distributed/test_distributed.py:2139]: it looks like with minor modifications to _prepare_dummy_data to handle variable batch size you should be able to reuse _test_DistributedDataParallel_SyncBatchNorm?

File: torch/nn/modules/_functions.py [hunk 2]
```diff
@@ -77,17 +75,18 @@ def backward(self, grad_output):
...
 
             # wait on the async communication to finish
-            mean_dy_all_reduce.wait()
-            mean_dy_xmu_all_reduce.wait()
+            sum_dy_all_reduce.wait()
+            sum_dy_xmu_all_reduce.wait()
 
-            mean_dy.div_(world_size)
-            mean_dy_xmu.div_(world_size)
+            dividor = count_tensor.sum()
```
Comment (ngimel) [inline torch/nn/modules/_functions.py:87]: divisor

**Review by ngimel — APPROVED** [review]
Looks good, I have minor comments.

File: test/distributed/test_distributed.py (hunk 1 above)
Comment (jjsjann123) [reply test/distributed/test_distributed.py:2139]: I can do so. I was worried about being too intrusive, as I would need to plumb the extra scaling factor all the way through `_test_DDP_5iter` as well (scale from this line https://github.com/pytorch/pytorch/pull/36382/files/e65860367bcb89ddde44c10922b1f9bd82417e67#diff-3469d9b41904e6ed24347065234034e2R2177)

But I changed `_test_DDP_HELP` already so I'll might do it as well to keep code simpler :man_shrugging:

File: torch/nn/modules/_functions.py [hunk 3]
```diff
@@ -46,9 +46,8 @@ def forward(self, input, weight, bias, running_mean, running_var, eps, momentum,
             count_all.view(-1).long().tolist()
         )
 
-        self.save_for_backward(input, weight, mean, invstd)
+        self.save_for_backward(input, weight, mean, invstd, count_all)
```
Comment (fmassa) [inline torch/nn/modules/_functions.py:49]: nit: the permissions of this file have changed

Comment (jjsjann123) [reply torch/nn/modules/_functions.py:49]: Good catch. Will undo the permission change.

**Pull request closed without merging**